### PR TITLE
NODE-1295 Implement a connection string parser for the core module

### PR DESCRIPTION
**Pull request by mbroadst**
This is a partial rewrite of the connection string parser from `native` ported to `core`. The impetus for this was a need to parse connection strings during the SDAM refactor, but it is also a feature oft-requested from users (in particular Mongoose). The parser does not concern itself with the knowledge of any particular key (outside of the single deprecated key in the connection string spec), but instead serves to do the basic work of connection string parsing, leaving such specific concerns to the porcelain layer.

It should be noted this is _another_ handcrafted parser, and it's my strong feel that we should instead develop a grammar for our particular flavor of URI that can be shared amongst mongo drivers. Having said that, I feel like the refactoring here improves and modernizes the original and should satisfy our needs in the field.

### CODE REVIEW

**Comment by daprahamian** [comment]
@mbroadst Are there any changes from the `node-mongodb-native` version? Or is this just moving the files?

**Comment by mbroadst** [comment]
@daprahamian yes, significant changes. The output is completely different, but now up to spec. It turns out we weren't even checking that the unit tests expected output matched the result of parsing previously!

File: lib/uri_parser.js [hunk 1]
```diff
@@ -0,0 +1,288 @@
...
+function parseSrvConnectionString(uri, options, callback) {
+  const result = URL.parse(uri);
+
+  if (result.hostname.split('.').length < 3) {
+    return callback(new Error('URI does not have hostname, domain name and tld'));
+  }
+
+  result.domainLength = result.hostname.split('.').length;
+  if (result.pathname && result.pathname.match(',')) {
+    return callback(new Error('Invalid URI, cannot contain multiple hostnames'));
```
[the remark this answers is not in the document]
Comment (mbroadst) [reply lib/uri_parser.js]: sure, `MongoParseError` ?

[the remark this answers is not in the document]
Comment (daprahamian) [reply lib/uri_parser.js]: Yeah, that sounds good.

File: lib/uri_parser.js [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+      connectionStringOptions.push('ssl=true');
+    }
+
+    // Keep original uri options
+    if (result.search) {
+      connectionStringOptions.push(result.search.replace('?', ''));
+    }
+
+    dns.resolveTxt(result.host, (err, record) => {
+      if (err && err.code !== 'ENODATA') return callback(err);
```
[the remark this answers is not in the document]
Comment (mbroadst) [reply lib/uri_parser.js]: 👍

File: lib/uri_parser.js [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+    let connectionStrings = addresses.map(
+      (address, i) =>
+        i === 0 ? `${base}${address.name}:${address.port}` : `${address.name}:${address.port}`
+    );
+
+    let connectionString = connectionStrings.join(',') + '/';
+    let connectionStringOptions = [];
+
+    // Default to SSL true
+    if (!options.ssl && !result.search) {
```
[the remark this answers is not in the document]
Comment (mbroadst) [reply lib/uri_parser.js]: sure, this section was actually a direct port of Jess' work earlier this month - wonder we didn't catch it then *shrug*

File: lib/uri_parser.js [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+      connectionStringOptions.push('ssl=true');
+    } else if (!options.ssl && result.search && !result.search.match('ssl')) {
```
[the remark this answers is not in the document]
Comment (jlord) [reply lib/uri_parser.js]: Ah, good point

[the remark this answers is not in the document]
Comment (mbroadst) [reply lib/uri_parser.js]: good call, we can easily have node's parser parse the query items and check that way

File: lib/uri_parser.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (mbroadst) [reply lib/uri_parser.js]: 👍

File: lib/uri_parser.js [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+  // special cases for known deprecated options
+  if (result.wtimeout && result.wtimeoutms) {
+    delete result.wtimeout;
+    console.warn('Unsupported option `wtimeout` specified');
+  }
+
+  return Object.keys(result).length ? result : null;
+}
+
+const SUPPORTED_PROTOCOLS = ['mongodb', 'mongodb+srv'];
```
[the remark this answers is not in the document]
Comment (mbroadst) [reply lib/uri_parser.js]: 👍

File: test/tests/spec/connection-string/valid-host_identifiers.json [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+          "type": "hostname",
+          "host": "umläut.example.com",
+          "port": null
+        }
+      ],
+      "auth": null,
+      "options": null
+    }
+  ]
+}
```
[the remark this answers is not in the document]
Comment (mbroadst) [reply test/tests/spec/connection-string/valid-host_identifiers.json]: 👍

File: lib/uri_parser.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (mbroadst) [reply lib/uri_parser.js]: 👍

File: lib/uri_parser.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (mbroadst) [reply lib/uri_parser.js]: 👍

**Comment by mbroadst** [comment]
@jlord @daprahamian gooood to go

**Pull request merged**